File: steps/typescript/recovery-monitor.step.ts
// steps/typescript/recovery-monitor.step.ts
import { EventConfig, Handlers } from 'motia';
import { TSStore } from './ts-store';

export const config = {
  type: 'event',
  name: 'TsRecoveryMonitor',
  description: 'Monitors pet recovery progress and schedules follow-up health checks',
  subscribes: ['ts.treatment.started', 'ts.treatment.completed'],
  emits: [],
  flows: ['TsPetManagement']
};

export const handler: Handlers['TsRecoveryMonitor'] = async (input, { logger }) => {
  const { petId, treatmentType, treatmentStatus } = input;

  if (logger) {
    logger.info('🩺 Recovery Monitor triggered', { petId, treatmentType, treatmentStatus });
  }

  try {
    const pet = TSStore.get(petId);
    if (!pet) {
      if (logger) {
        logger.error('❌ Pet not found for recovery monitoring', { petId });
      }
      return;
    }

    if (treatmentStatus === 'started') {
      // Treatment just started - set up monitoring
      const recoveryPlan = {
        petId,
        treatmentType,
        startedAt: Date.now(),
        expectedRecoveryTime: getExpectedRecoveryTime(treatmentType),
        monitoringSchedule: generateMonitoringSchedule(treatmentType),
        milestones: generateRecoveryMilestones(treatmentType),
        currentPhase: 'initial_treatment'
      };

      if (logger) {
        logger.info('📋 Recovery plan created', {
          petId,
          treatmentType,
          expectedRecoveryTime: recoveryPlan.expectedRecoveryTime,
          milestones: recoveryPlan.milestones.length
        });
      }

      // Recovery plan created successfully (no emit - no subscribers)

    } else if (treatmentStatus === 'completed') {
      // Treatment completed - schedule follow-up
      const followUpSchedule = {
        petId,
        treatmentCompletedAt: Date.now(),
        followUpChecks: [
          { type: 'immediate', scheduledAt: Date.now() + (2 * 60 * 60 * 1000) }, // 2 hours
          { type: 'daily', scheduledAt: Date.now() + (24 * 60 * 60 * 1000) }, // 1 day
          { type: 'weekly', scheduledAt: Date.now() + (7 * 24 * 60 * 60 * 1000) } // 1 week
        ],
        recoveryIndicators: getRecoveryIndicators(treatmentType),
        readyForDischarge: false
      };

      if (logger) {
        logger.info('✅ Treatment completed, scheduling follow-up', {
          petId,
          followUpChecks: followUpSchedule.followUpChecks.length
        });
      }

      // Follow-up scheduled successfully (no emit - no subscribers)
    }

  } catch (error: any) {
    if (logger) {
      logger.error('❌ Recovery monitoring failed', { petId, error: error.message });
    }
  }
};

function getExpectedRecoveryTime(treatmentType: string): string {
  switch (treatmentType.toLowerCase()) {
    case 'emergency surgery': return '2-4 weeks';
    case 'respiratory treatment': return '1-2 weeks';
    case 'pain management': return '3-7 days';
    case 'antibiotic treatment': return '7-14 days';
    case 'fever management': return '3-5 days';
    default: return '1-2 weeks';
  }
}

function generateMonitoringSchedule(treatmentType: string): any[] {
  const baseSchedule = [
    { time: 'every 2 hours', check: 'vital signs', priority: 'high' },
    { time: 'every 6 hours', check: 'medication compliance', priority: 'high' },
    { time: 'daily', check: 'wound healing', priority: 'medium' },
    { time: 'daily', check: 'appetite and hydration', priority: 'medium' }
  ];

  if (treatmentType.toLowerCase().includes('surgery')) {
    baseSchedule.push(
      { time: 'every 4 hours', check: 'incision site', priority: 'high' },
      { time: 'daily', check: 'mobility assessment', priority: 'medium' }
    );
  }

  if (treatmentType.toLowerCase().includes('respiratory')) {
    baseSchedule.push(
      { time: 'every 3 hours', check: 'breathing pattern', priority: 'high' },
      { time: 'daily', check: 'oxygen levels', priority: 'high' }
    );
  }

  return baseSchedule;
}

function generateRecoveryMilestones(treatmentType: string): any[] {
  const milestones = [
    { day: 1, milestone: 'Initial treatment response', status: 'pending' },
    { day: 3, milestone: 'Pain management effectiveness', status: 'pending' },
    { day: 7, milestone: 'Primary healing indicators', status: 'pending' },
    { day: 14, milestone: 'Full recovery assessment', status: 'pending' }
  ];

  if (treatmentType.toLowerCase().includes('surgery')) {
    milestones.push(
      { day: 2, milestone: 'Incision healing check', status: 'pending' },
      { day: 10, milestone: 'Stitch removal readiness', status: 'pending' }
    );
  }

  return milestones;
}

function getRecoveryIndicators(treatmentType: string): string[] {
  const baseIndicators = [
    'Normal appetite',
    'Active behavior',
    'No signs of pain',
    'Normal vital signs'
  ];

  if (treatmentType.toLowerCase().includes('surgery')) {
    baseIndicators.push('Incision healing well', 'No signs of infection', 'Good mobility');
  }

  if (treatmentType.toLowerCase().includes('respiratory')) {
    baseIndicators.push('Normal breathing pattern', 'Good oxygen saturation', 'No coughing');
  }

  return baseIndicators;
}

File: steps/python/recovery_monitor_step.py
# steps/python/recovery_monitor.step.py
import sys
import os
import time

# Add parent directory to path for imports
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
from services import pet_store

config = {
    "type": "event",
    "name": "PyRecoveryMonitor",
    "description": "Monitors pet recovery progress and schedules follow-up health checks",
    "subscribes": ["py.treatment.started", "py.treatment.completed"],
    "emits": [],
    "flows": ["PyPetManagement"]
}

def get_expected_recovery_time(treatment_type):
    treatment_lower = treatment_type.lower()
    if 'emergency surgery' in treatment_lower:
        return '2-4 weeks'
    elif 'respiratory treatment' in treatment_lower:
        return '1-2 weeks'
    elif 'pain management' in treatment_lower:
        return '3-7 days'
    elif 'antibiotic treatment' in treatment_lower:
        return '7-14 days'
    elif 'fever management' in treatment_lower:
        return '3-5 days'
    else:
        return '1-2 weeks'

def generate_monitoring_schedule(treatment_type):
    base_schedule = [
        {'time': 'every 2 hours', 'check': 'vital signs', 'priority': 'high'},
        {'time': 'every 6 hours', 'check': 'medication compliance', 'priority': 'high'},
        {'time': 'daily', 'check': 'wound healing', 'priority': 'medium'},
        {'time': 'daily', 'check': 'appetite and hydration', 'priority': 'medium'}
    ]

    treatment_lower = treatment_type.lower()
    if 'surgery' in treatment_lower:
        base_schedule.extend([
            {'time': 'every 4 hours', 'check': 'incision site', 'priority': 'high'},
            {'time': 'daily', 'check': 'mobility assessment', 'priority': 'medium'}
        ])

    if 'respiratory' in treatment_lower:
        base_schedule.extend([
            {'time': 'every 3 hours', 'check': 'breathing pattern', 'priority': 'high'},
            {'time': 'daily', 'check': 'oxygen levels', 'priority': 'high'}
        ])

    return base_schedule

def generate_recovery_milestones(treatment_type):
    milestones = [
        {'day': 1, 'milestone': 'Initial treatment response', 'status': 'pending'},
        {'day': 3, 'milestone': 'Pain management effectiveness', 'status': 'pending'},
        {'day': 7, 'milestone': 'Primary healing indicators', 'status': 'pending'},
        {'day': 14, 'milestone': 'Full recovery assessment', 'status': 'pending'}
    ]

    treatment_lower = treatment_type.lower()
    if 'surgery' in treatment_lower:
        milestones.extend([
            {'day': 2, 'milestone': 'Incision healing check', 'status': 'pending'},
            {'day': 10, 'milestone': 'Stitch removal readiness', 'status': 'pending'}
        ])

    return milestones

def get_recovery_indicators(treatment_type):
    base_indicators = [
        'Normal appetite',
        'Active behavior',
        'No signs of pain',
        'Normal vital signs'
    ]

    treatment_lower = treatment_type.lower()
    if 'surgery' in treatment_lower:
        base_indicators.extend(['Incision healing well', 'No signs of infection', 'Good mobility'])

    if 'respiratory' in treatment_lower:
        base_indicators.extend(['Normal breathing pattern', 'Good oxygen saturation', 'No coughing'])

    return base_indicators

async def handler(input_data, ctx=None):
    logger = getattr(ctx, 'logger', None) if ctx else None
    emit = getattr(ctx, 'emit', None) if ctx else None
    
    pet_id = input_data.get('petId')
    treatment_type = input_data.get('treatmentType', 'general')
    treatment_status = input_data.get('treatmentStatus')

    if logger:
        logger.info('🩺 Recovery Monitor triggered', {'petId': pet_id, 'treatmentType': treatment_type, 'treatmentStatus': treatment_status})

    try:
        pet = pet_store.get(pet_id)
        if not pet:
            if logger:
                logger.error('❌ Pet not found for recovery monitoring', {'petId': pet_id})
            return

        if treatment_status == 'started':
            # Treatment just started - set up monitoring
            recovery_plan = {
                'petId': pet_id,
                'treatmentType': treatment_type,
                'startedAt': int(time.time() * 1000),
                'expectedRecoveryTime': get_expected_recovery_time(treatment_type),
                'monitoringSchedule': generate_monitoring_schedule(treatment_type),
                'milestones': generate_recovery_milestones(treatment_type),
                'currentPhase': 'initial_treatment'
            }

            if logger:
                logger.info('📋 Recovery plan created', {
                    'petId': pet_id,
                    'treatmentType': treatment_type,
                    'expectedRecoveryTime': recovery_plan['expectedRecoveryTime'],
                    'milestones': len(recovery_plan['milestones'])
                })

            if emit:
                await emit({
                    'topic': 'py.recovery.progress',
                    'data': {
                        'petId': pet_id,
                        'recoveryPlan': recovery_plan,
                        'nextSteps': [
                            'Begin treatment monitoring',
                            'Schedule daily health checks',
                            'Update medication schedule',
                            'Notify staff of special care requirements'
                        ],
                        'timestamp': int(time.time() * 1000)
                    }
                })

        elif treatment_status == 'completed':
            # Treatment completed - schedule follow-up
            follow_up_schedule = {
                'petId': pet_id,
                'treatmentCompletedAt': int(time.time() * 1000),
                'followUpChecks': [
                    {'type': 'immediate', 'scheduledAt': int(time.time() * 1000) + (2 * 60 * 60 * 1000)},  # 2 hours
                    {'type': 'daily', 'scheduledAt': int(time.time() * 1000) + (24 * 60 * 60 * 1000)},  # 1 day
                    {'type': 'weekly', 'scheduledAt': int(time.time() * 1000) + (7 * 24 * 60 * 60 * 1000)}  # 1 week
                ],
                'recoveryIndicators': get_recovery_indicators(treatment_type),
                'readyForDischarge': False
            }

            if logger:
                logger.info('✅ Treatment completed, scheduling follow-up', {
                    'petId': pet_id,
                    'followUpChecks': len(follow_up_schedule['followUpChecks'])
                })

            if emit:
                await emit({
                    'topic': 'py.health.check.scheduled',
                    'data': {
                        'petId': pet_id,
                        'followUpSchedule': follow_up_schedule,
                        'nextSteps': [
                            'Monitor recovery indicators',
                            'Schedule follow-up appointments',
                            'Prepare discharge paperwork',
                            'Update pet status when ready'
                        ],
                        'timestamp': int(time.time() * 1000)
                    }
                })

    except Exception as error:
        if logger:
            logger.error('❌ Recovery monitoring failed', {'petId': pet_id, 'error': str(error)})
